File: 알고리즘 문제풀이/kakao/괄호변환.py
#균형 잡힌 괄호: ( 와 )의 갯수가 같음
from collections import defaultdict
def f(s1): #올바른 괄호 체크
    st,f=[],1
    for s in s1:
        if s=='(':
            st.append(s)
        else:
            if len(st)>0:
                st.pop(0)
            else:
                return 0
    if len(st)>0:
        return 0
    else:
        return 1
def solution(p):
    answer = ''
    if len(p)>0: #1
        if f(p):
            return p
        else: 
            d=defaultdict(int) #2
            s1,s2="",""
            for i in range(len(p)):
                d[p[i]]+=1
                if d['(']==d[')']:
                    s1=p[:i+1]
                    s2=p[i+1:]
                    break
            #3 u가 "올바른 괄호 문자열"일 경우
            if(f(s1)):
                answer+=s1
                tmp=solution(s2)
                answer+=tmp #3-1
            #4 u가 "올바른 괄호 문자열"이 아닐경우
            else:
                #4-1~4-3
                tmp1="("
                tmp1+=solution(s2)
                tmp1+=')'
                #4-4
                for i in range(1,len(s1)-1):
                    if(s1[i]=='('):
                        tmp1+=')'
                    else:
                        tmp1+='('
                #4-5
                answer=tmp1
    return answer

File: 카카오 코딩 문제/괄호 변환.js
const solution = (p) => {
  let answer = "";
  if (p.length === 0) return "";
  const check = (w) => {
    let u = "",
      v = "",
      tmp = "",
      st = [],
      le = 0,
      ri = 0;
    if (w.length > 0) {
      for (let i = 0; i < w.length; i++) {
        if (w[i] === "(") le += 1;
        else if (w[i] === ")") ri += 1;
        if (le === ri) {
          u = w.slice(0, i + 1);
          v = w.slice(i + 1, w.length);
          break;
        }
      }
      //u가 올바른 괄호인지 체크
      let flag = true;
      for (const v of u) {
        if (v === "(") st.push("(");
        else {
          if (st.length > 0) st.pop(")");
          else {
            flag = false;
            break;
          }
        }
      }
      if (flag === false) {
        //균형 잡힌 괄호인 경우
        return (
          "(" +
          check(v) +
          ")" +
          u
            .slice(1, u.length - 1)
            .split("")
            .map((item) => (item === ")" ? "(" : ")"))
            .join("")
        );
      } else {
        //올바른 괄호인 경우
        return u + check(v);
      }
    }
    return "";
  };
  answer = check(p);
  return answer;
};
solution("(()())()");
console.log("-----");
solution(")(");
console.log("-----");
solution("()))((()");
